models/user_setting.go

```diff
@@ -0,0 +1,116 @@
+}
+
+// BeforeUpdate will be invoked by XORM before updating a record
+func (userSetting *UserSetting) BeforeUpdate() {
+	userSetting.Key = strings.ToLower(userSetting.Key)
+}
+
+// BeforeDelete will be invoked by XORM before updating a record
+func (userSetting *UserSetting) BeforeDelete() {
+	userSetting.Key = strings.ToLower(userSetting.Key)
+}
+
+func init() {
+	db.RegisterModel(new(UserSetting))
+}
+
+// GetUserSetting returns specific settings from user
+func GetUserSetting(uid int64, keys []string) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, len(keys))
+	if err := db.GetEngine(db.DefaultContext).
+		Where("user_id=?", uid).
+		And(builder.In("key", keys)).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// GetAllUserSettings returns all settings from user
+func GetAllUserSettings(uid int64) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
+	if err := db.GetEngine(db.DefaultContext).
+		Where("user_id=?", uid).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// DeleteUserSetting deletes a specific setting for a user
+func DeleteUserSetting(userSetting *UserSetting) error {
+	sess := db.NewSession(db.DefaultContext)
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if _, err := sess.Delete(userSetting); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// SetUserSetting updates a users' setting for a specific key
+func SetUserSetting(userSetting *UserSetting) error {
+	return upsertUserSettingValue(db.GetEngine(db.DefaultContext), userSetting.UserID, userSetting.Key, userSetting.Value)
+}
+
+func upsertUserSettingValue(e db.Engine, userID int64, key string, value string) (err error) {
+	// Intentionally lowercase key here as XORM may not pick it up via Before* actions
+	key = strings.ToLower(key)
+	// An atomic UPSERT operation (INSERT/UPDATE) is the only operation
+	// that ensures that the key is actually locked.
+	switch {
+	case setting.Database.UseSQLite3 || setting.Database.UsePostgreSQL:
+		_, err = e.Exec("INSERT INTO `user_setting` (user_id, key, value) "+
+			"VALUES (?,?,?) ON CONFLICT (user_id,key) DO UPDATE SET value = ?",
+			userID, key, value, value)
+	case setting.Database.UseMySQL:
+		_, err = e.Exec("INSERT INTO `user_setting` (user_id, key, value) "+
+			"VALUES (?,?,?) ON DUPLICATE KEY UPDATE value = ?",
+			userID, key, value, value)
+	case setting.Database.UseMSSQL:
+		// https://weblogs.sqlteam.com/dang/2009/01/31/upsert-race-condition-with-merge/
+		_, err = e.Exec("MERGE `user_setting` WITH (HOLDLOCK) as target "+
+			"USING (SELECT ? AS user_id, ? AS key) AS src "+
+			"ON src.user_id = target.user_id AND src.key = target.key "+
+			"WHEN MATCHED THEN UPDATE SET target.value = ? "+
+			"WHEN NOT MATCHED THEN INSERT (user_id, key, value) "+
+			"VALUES (src.user_id, src.key, ?);",
+			userID, key, value, value)
```

Comment:
I previously had what you were suggesting, but due to potential conflicts happening it was suggested to keep this logic in the DB directly so that DB native upsert logic could be used.

cc: @zeripath who made this suggestion.

Reply:
"potential conflicts": what are the conflicts?

Reply:
If two operations happen simultaneously one would fail, as both would attempt an update, and when that fails both would attempt an insert where only one would be successful. By using built in upsert then both operations would be successful

Reply:
But that's the correct behavior.  

Assume the backend gets two ﻿﻿simultaneous request `set A=1`, `set A=2`, they come at the same time, we do not know which one is really the first one. And when the `save` request reaches database, we assume `A=1` is a little earlier than `A=2`:

With your code: first A is inserted to 1, then A is updated to 2.

With my code: the first update changes 0 row and we get a txn lock, and then do a insert to 1. the second update is blocked by the first update lock, when the lock is released, the second update is executed then A is still updated to 2, then one row is changed and no more insert.